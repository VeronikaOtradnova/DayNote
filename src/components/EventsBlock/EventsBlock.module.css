.list::-webkit-scrollbar {
  width: 5px;
}

.list::-webkit-scrollbar-track {
  background: #f1f1f1;
  border-radius: 10px;
}

.list::-webkit-scrollbar-thumb {
  background: #888;
  border-radius: 10px;
}

.list::-webkit-scrollbar-thumb:hover {
  background: #555;
}

@media (min-width: 1000px) {
  .wrapper {
    width: 60%;
    padding-left: 15px;
  }

  .list {
    max-height: 350px;
    padding-right: 10px;
    overflow: auto;
  }
}