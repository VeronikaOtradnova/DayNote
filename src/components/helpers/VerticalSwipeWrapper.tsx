import { useState } from "react"

interface IProps {
  children: React.ReactNode,
  onTop?: () => void,
  onBottom?: () => void,
}

export function VerticalSwipeWrapper({children, onTop = () => {}, onBottom = () => {}}: IProps) {
  const [touchStart, setTouchStart] = useState<number | null>(null)
  const [touchEnd, setTouchEnd] = useState<number | null>(null)

  // the required distance between touchStart and touchEnd to be detected as a swipe
  const minSwipeDistance = 50

  const onTouchStart = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(null)
    setTouchStart(e.targetTouches[0].clientY)
  }

  const onTouchMove = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(e.targetTouches[0].clientY);
  };

  const onTouchEnd = () => {
    if (!touchStart || !touchEnd) return
    const distance = touchStart - touchEnd
    const isTopSwipe = distance > minSwipeDistance
    const isBottomSwipe = distance < -minSwipeDistance
    if (isTopSwipe || isBottomSwipe) {
      isTopSwipe ? onTop() : onBottom()
    }
  }

  return (
    <div 
      onTouchStart={onTouchStart} 
      onTouchMove={onTouchMove} 
      onTouchEnd={onTouchEnd} 
    >
      {children}
    </div>
  )
}