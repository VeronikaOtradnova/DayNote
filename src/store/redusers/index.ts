import { combineReducers } from "redux";
import { dayReducer } from "./dayReducer";
import { calendarReducer } from "./calendarReducer";

export const rootReducer = combineReducers({
  day: dayReducer,
  calendar: calendarReducer
})

export type TRootState = ReturnType<typeof rootReducer>