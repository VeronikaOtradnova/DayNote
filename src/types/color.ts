export enum colors {
  none = 'none',
  green = 'green',
  pink = 'pink',
  gray = 'gray',
  blue = 'blue',
}

export type TColor = colors.green | colors.pink | colors.gray | colors.none | colors.blue;