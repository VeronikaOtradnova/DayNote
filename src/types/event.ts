export enum eventPriorities {
  EXTRA_HIGH = 'очень высокий',
  HIGH = 'высокий',
  MEDIUM = 'средний',
  LOW = 'низкий',
}

export interface IEvent {
  id: string,
  day: number,
  time: number,
  title: string,
  description: string,
  priority: eventPriorities,
  selected: boolean,
}

export enum eventActionTypes {
  ADD_EVENT = 'ADD_EVENT',
  EDIT_EVENT = 'EDIT_EVENT',
  REMOVE_EVENTS = 'REMOVE_EVENTS'
}

export interface IAddEventAction {
  type: eventActionTypes.ADD_EVENT,
  payload: IEvent,
}
export interface IEditEventAction {
  type: eventActionTypes.EDIT_EVENT,
  payload: IEvent,
}
export interface IRemoveEventsAction {
  type: eventActionTypes.REMOVE_EVENTS,
  payload: string[],
}

export type TEventAction = IAddEventAction | IEditEventAction | IRemoveEventsAction;

export interface IEventState {
  events: IEvent[]
}