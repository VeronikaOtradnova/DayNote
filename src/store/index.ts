import { applyMiddleware, createStore } from "redux";
import { rootReducer } from "./redusers";
import { thunk } from "redux-thunk";

const stateFromStorage = localStorage.getItem('reduxState')
export const persistedState = stateFromStorage ? JSON.parse(stateFromStorage): {}

export const store = createStore(rootReducer, persistedState, applyMiddleware(thunk));

export type AppDispatch = typeof store.dispatch;

store.subscribe(()=>{
  localStorage.setItem('reduxState', JSON.stringify(store.getState()))
})