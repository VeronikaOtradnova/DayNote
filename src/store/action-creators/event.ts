import { IAddEventAction, IEditEventAction, IEvent, IRemoveEventsAction, eventActionTypes } from "../../types/event";

export const addEvent = (event: IEvent):IAddEventAction => ({
  type: eventActionTypes.ADD_EVENT,
  payload: event
})

export const editEvent = (event: IEvent):IEditEventAction => ({
  type: eventActionTypes.EDIT_EVENT,
  payload: event
})

export const removeEvents = (ids: string[]):IRemoveEventsAction => ({
  type: eventActionTypes.REMOVE_EVENTS,
  payload: ids
})