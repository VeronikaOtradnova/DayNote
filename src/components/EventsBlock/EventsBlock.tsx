import { useState } from 'react';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import styles from './EventsBlock.module.css';
import { PlusBtn } from '../generic/buttons/PlusBtn/PlusBtn';
import { NewEventForm } from './forms/NewEventForm/NewEventForm';
import { XBtn } from '../generic/buttons/XBtn/XBtn';
import { EventItem } from './EventItem/EventItem';
import { IEvent } from '../../types/event';
import { EditEventBtn } from './btns/EditEventBtn/EditEventBtn';
import { RemoveEventsBtn } from './btns/RemoveEventsBtn/RemoveEventsBtn';
import { XSquareBtn } from '../generic/buttons/XSquareBtn/XSquareBtn';

export function EventsBlock() {
  const {events} = useTypedSelector(store => store.event);
  const {currentDay} = useTypedSelector(store => store.day);
  const eventsList = events.filter(event => event.day === currentDay);
  const checkedItems = eventsList.filter(event => event.selected);
  const [isFormOpen, setFormOpen] = useState(false);
  
  return (
    <>
      <div className={`block-wrapper ${styles.wrapper}`}>
        <div className='block-title-wrapper'>
          <h2 className='block-title'>События</h2>

          {
            checkedItems.length === 1 && 
            <EditEventBtn />
          }

          {
            checkedItems.length > 0 && 
            <RemoveEventsBtn />
          }

          {
            isFormOpen ?
            <XSquareBtn 
              onClick={() => setFormOpen(false)}
              testId='events-block__hide-form-btn'
              bgSize={30}
            />
            :
            <PlusBtn
              onClick={() => setFormOpen(true)}
              testId='events-block__show-form-btn'
              bgSize={30}
            />
          }
        </div>

        {
          isFormOpen &&
          
            <NewEventForm
              hideForm={() => setFormOpen(false)}
            />
          
        }

        <ul className={`list ${styles.list}`}>
          {
            eventsList && eventsList.sort((a: IEvent, b: IEvent) => b.time < a.time ? 1 : -1).map(event => <EventItem event={event} key={event.id} />)
          }
        </ul>
      </div>
    </>
  )
}