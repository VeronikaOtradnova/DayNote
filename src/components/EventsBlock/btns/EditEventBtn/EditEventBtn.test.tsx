import { screen } from "@testing-library/react"
import { generateString } from "../../../../helpers/generateString"
import { TRootState } from "../../../../store/redusers"
import { renderWithRedux, testInitialState } from "../../../../tests/helpers/renderWithRedux"
import { colors } from "../../../../types/color"
import { eventPriorities } from "../../../../types/event"
import { EventsBlock } from "../../EventsBlock"
import userEvent from "@testing-library/user-event"
import { act } from "react-dom/test-utils"

const zeroSelectedEventsInitialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
    }]
  },
  event: {
    events: [
      {
        day: +(new Date(2020, 0, 1)),
        description: 'test1 description',
        id: generateString(),
        priority: eventPriorities.LOW,
        selected: false,
        time: 12 * 3600000 + 0 * 60000,
        title: 'test1'
      },
      {
        day: +(new Date(2020, 0, 1)),
        description: 'test2 description',
        id: generateString(),
        priority: eventPriorities.MEDIUM,
        selected: false,
        time: 14 * 3600000 + 30 * 60000,
        title: 'test2'
      }
    ]
  }
}

const oneSelectedEventInitialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
    }]
  },
  event: {
    events: [
      {
        day: +(new Date(2020, 0, 1)),
        description: 'test1 description',
        id: generateString(),
        priority: eventPriorities.LOW,
        selected: true,
        time: 12 * 3600000 + 0 * 60000,
        title: 'test1'
      },
      {
        day: +(new Date(2020, 0, 1)),
        description: 'test2 description',
        id: generateString(),
        priority: eventPriorities.MEDIUM,
        selected: false,
        time: 14 * 3600000 + 30 * 60000,
        title: 'test2'
      }
    ]
  }
}

const twoSelectedEventsInitialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
    }]
  },
  event: {
    events: [
      {
        day: +(new Date(2020, 0, 1)),
        description: 'test1 description',
        id: generateString(),
        priority: eventPriorities.LOW,
        selected: true,
        time: 12 * 3600000 + 0 * 60000,
        title: 'test1'
      },
      {
        day: +(new Date(2020, 0, 1)),
        description: 'test2 description',
        id: generateString(),
        priority: eventPriorities.MEDIUM,
        selected: true,
        time: 14 * 3600000 + 30 * 60000,
        title: 'test2'
      }
    ]
  }
}

describe('EDIT-EVENT-BTN-TESTING', () => {
  test('Should render btn when one event is selected', () => {
    renderWithRedux(<EventsBlock />, {initialState: oneSelectedEventInitialState});

    expect(screen.getByTestId('edit-event-btn')).toBeInTheDocument();
    expect(screen.getByTestId('edit-event-btn')).toMatchSnapshot();
  });

  test('Should not render btn when no selected events', () => {
    renderWithRedux(<EventsBlock />, {initialState: zeroSelectedEventsInitialState});
    
    expect(screen.queryByTestId('edit-event-btn')).not.toBeInTheDocument();
  });

  test('Should not render btn when more than one events are selected', () => {
    renderWithRedux(<EventsBlock />, {initialState: twoSelectedEventsInitialState});
    
    expect(screen.queryByTestId('edit-event-btn')).not.toBeInTheDocument();
  });

  test('Should render edit-modal after clicking the btn', async () => {
    const user = userEvent.setup();

    renderWithRedux(<EventsBlock />, {initialState: oneSelectedEventInitialState});
    const btn = screen.getByTestId('edit-event-btn');

    await act(async () => {
      await user.click(btn);
    })

    expect(screen.getByTestId('edit-event-modal')).toBeInTheDocument();
  })
})