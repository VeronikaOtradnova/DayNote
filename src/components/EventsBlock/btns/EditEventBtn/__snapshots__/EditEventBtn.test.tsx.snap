// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EDIT-EVENT-BTN-TESTING Should render btn when one event is selected 1`] = `
<button
  class="btn"
  data-testid="edit-event-btn"
/>
`;
